### PR TITLE
make auxiliary heads in pretrained models optional

### PR DESCRIPTION
**Pull request by Separius**
it is related to pytorch/pytorch#18668

### CODE REVIEW

File: torchvision/models/googlenet.py [hunk 1]
```diff
@@ -62,8 +69,9 @@ def __init__(self, num_classes=1000, aux_logits=True, transform_input=False, ini
         self.inception5a = Inception(832, 256, 160, 320, 32, 128, 128)
         self.inception5b = Inception(832, 384, 192, 384, 48, 128, 128)
 
-        self.aux1 = InceptionAux(512, num_classes)
-        self.aux2 = InceptionAux(528, num_classes)
+        if aux_logits:
```
Comment (fmassa) [inline torchvision/models/googlenet.py]: This partially reverts https://github.com/pytorch/vision/pull/821 from @TheCodez 
I've made a comment in https://github.com/pytorch/vision/pull/821#discussion_r270351338 which is related to this change, and I think that @TheCodez reasoning was good.

File: torchvision/models/inception.py [hunk 2]
```diff
@@ -29,8 +29,17 @@ def inception_v3(pretrained=False, **kwargs):
     if pretrained:
         if 'transform_input' not in kwargs:
             kwargs['transform_input'] = True
+        if 'aux_logits' in kwargs:
+            original_aux_logits = kwargs['aux_logits']
+            kwargs['aux_logits'] = True
+        else:
+            original_aux_logits = True
```
[the remark this answers is not in the document]
Comment (Separius) [reply torchvision/models/inception.py]: yes, I assumed that the pretrained models all have `aux_logits` and they are trained but based on #821 it seems that they are actually not trained(but included in the pretrained models)

so doesn't it make sense to disable aux_heads in the pretrained models and update the pretrained models(.pth files)?

it breaks backward compatibility though and maybe it makes sense to set `strict` to False in `load_from_state`?

File: torchvision/models/googlenet.py (hunk 1 above)
Comment (Separius) [reply torchvision/models/googlenet.py]: the reasoning makes sense. so should we do something similar in the inception model?

File: torchvision/models/inception.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/models/inception.py]: So, one of my ideas was to completely disable `.train(True)` mode, and not have the aux branches at all.
Did you train a checkpoint on top of inception v3 without the aux heads?

[the remark this answers is not in the document]
Comment (Separius) [reply torchvision/models/inception.py]: completely disabling train mode(for pretrained models) is a bit harsh, don't you think?
I guess most of the time people freeze the net, train just the fc layer and then unfreeze the whole thing with a small lr. disabling train will break those codes.

No no, by reading #821 I thought that maybe aux paths are not trained at all in the `.pth` file and that's the reason they are discarded, am I mistaken?

[the remark this answers is not in the document]
Comment (TheCodez) [reply torchvision/models/inception.py]: @Separius only the aux heads of GoogLeNet don't have pretrained weights, the aux branch of inception v3 however is trained.

[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/models/inception.py]: In this case, I think we should leave GoogleNet as is, and then this is good to be merged?

[the remark this answers is not in the document]
Comment (Separius) [reply torchvision/models/inception.py]: @TheCodez so I guess we have two options then, either delete aux branch after loading it in the inception model(like my code) or do not allow `aux_logits=False` in the `load_pretrained` function

**Comment by TheCodez** [comment]
My question is why should we delete the aux branch for inception v3 but not for GoogleNet?
I agree with @fmassa that we should always create those branches and then just disable them if not desired.

@Separius code with the lines `del model.AuxLogits` and `if aux_logits:` removed should be fine if I'm not mistaken?
But we should double check that to make sure we don't break any existing models.

**Comment by Separius** [comment]
ok based on our discussion, here is what I propose:

both pretrained models should accept `aux_logits=True`(and False) and both models should treat it similarly(right now when `aux_logits` is false in the inception we don't instantiate those heads but we do it in the googlenet) **but** we must warn the user that using `aux_logits=True` in googlenet is not recommended(because they are not pretrained).

it won't break BC and it won't take extra space for the unused parameters and more importantly it allows fine-tuning, either with or without auxiliary heads.

what do you think?, @TheCodez @fmassa 

@TheCodez ragarding breaking existing models, right now `load_pretraind_inception` with  `aux_logits=True` fails, so I think it's safe to say that we won't break anything.

**Comment by TheCodez** [comment]
@Separius your reasoning seems good 👍 

> but we must warn the user that using aux_logits=True in googlenet is not recommended(because they are not pretrained).

This is important to document that they aren't pretrained.

> @TheCodez ragarding breaking existing models, right now load_pretraind_inception with aux_logits=True fails, so I think it's safe to say that we won't break anything.

Yeah my idea of always creating the aux branch for inception v3 as well would cause bc issues with finetuned models. 
So it's probably a good idea to be consistent in both googlenet and inception.

**Comment by Separius** [comment]
@TheCodez I also changed the ordering of the returned values from googlenet when aux is on to make it consistent with inception, I know it's a breaking change but I believe googlenet is not in the latest torchvision release, so we are safe to change it, right?

**Comment by TheCodez** [comment]
@Separius I think that’s fine.
What about ‘GoogLeNetOuputs’?

**Comment by Separius** [comment]
should I also add some tests to the `test_models.py`?
I mean, is it ok for the tests to download pretrained models? (I live in Iran and downloading all the pretrained models takes a lot of time. that's why I'm asking :)))

**Comment by fmassa** [comment]
@Separius let's not download pre-trained models in the tests. This can make the tests flaky due to connection errors
